### DISCUSSION TITLE
What gonna happen if I disable mkstemp

### DISCUSSION THREAD

**Discussion by DCjanus** [ImageMagick/ImageMagick#2783] [Help]
Hi, we are using imagemagick-core and imagemagick++ in our image process server, and we noticed that sometimes imagemagick call `mkstemp`, which create temp files and dentry cache for those file, but sometimes, core wouldn't GC those dentry cache, we don't know why yet.
In our core version and Docker version, mkstemp caused Slab memory leak, restart container is not working for us, is that OK if I want to disable mkstemp?

**Comment by dlemstra** [comment]
If you disable `mkstemp` some images will be unable to be converted because they require a file on disk. Another reason could be that you are running out of memory. ImageMagick then creates files on disk to store the pixel memory. I would advise you to not disable this but to investigate what is causing this. It might be possible that we leave a file on disk due to a bug. We also fixed some of those bugs a couple years ago. Are you running a recent version of ImageMagick?

**Reply by DCjanus** [reply]
Maybe I didn't explain clearly: We have a long-running image service, power by ImageMagick(thanks), and I found [those line](https://github.com/ImageMagick/ImageMagick/blob/master/MagickCore/random.c#L454-L479), they generate random seed via `mkstemp`. Temp file were removed, but stil caused some dentry cache, and sometimes, Linux kernel wouldn't recycle those dentry cache in time.

For now, we forked ImageMagick and remove those lines from it. This is really tricky, and if there is a better way to make sure ImageMagick would create files, that's would be really great for our long-running service.

**Comment by urban-warrior** [comment]
Both mkstemp() calls can be safety disabled.  For temporary files, there is a backup mechanism to create them if mkstemp() is not called or fails.  For random #'s, mkstemp() adds entropy but without the call there is still plenty of entropy to generate the random #  .

ImageMagick still requires temporary files, without mkstemp(), we call open() on a temporary file path instead.  You can of course set the temporary file path, it defaults to /tmp, but the temporary files are still created.  Are you suggesting that the dentry cache is affected by mkstemp() or creating temporary files in general?  Is the issue related to exclusive access to prevent race conditions for temporary files?  

To disable the mkstemp() call,  uncomment MAGICKCORE_HAVE_MKSTEMP in *baseconfig.h and rebuild/reinstall ImageMagick to eliminate the mkstemp() call.

**Reply by DCjanus** [reply]
Thanks your reply.

Recently, our image process service received some request to set transparency, and suddenly, it cost a lot of `Slab memory`(found this in `/proc/meminfo`), `slabtop` told us that most memory is in `dentry`. Remove those type requests, there is no more `Slab memory` leak.

In our code, we impl this feature via  `EvaluateImage(MultiplyEvaluateOperator)`, we gdb attach our process and found there is a lot `open` syscall, from [here](https://github.com/ImageMagick/ImageMagick/blob/d873b82c882de994c733e7320aa82d6393f2c1e6/MagickCore/random.c#L454-L479).

**Comment by urban-warrior** [comment]
Creating a temporary file is likely excessive work for generating entropy.  We'll replace this code with a new source of entropy within the next few days.

**Reply by DCjanus** [reply]
Thanks, actual, I think `server usage` is not main case to ImageMagick, that why I wasn't tring to remove those lines from your code.  But, Thanks.
